### DISCUSSION TITLE
How do you open the window switcher without alt+tab?

### DISCUSSION THREAD

**Discussion by erikbaender** [BuddiesOfBudgie/budgie-desktop#33] [Q&A]
I'm trying to create a script that will launch budgie's window switcher and hide all unfocused windows after it exits, but I can't get it to reliably open the window switcher.

I've tried these two approaches with xdotool, but both of them are extremely janky and only trigger the window switcher very vary rarely.

`xdotool key alt+Tab`
`xdotool keydown alt key Tab; sleep 2; xdotool keyup alt`

Most of the time both commands produce don't trigger the switcher and instead write this cryptic output to the terminal:

`^[	%      `

I'd like to avoid this janky way of emulating pressing alt and tab if possible, but I couldn't find any way to directly launch the window switcher. It doesn't seem to have its own executable, and the manpage for wmctrl also doesn't say anything about a command for the window switcher.

**Comment by guillotjulien** [comment]
To my knowledge, the tab switcher cannot be invoked individially without some code. Technically, you could open it via DBus (org.budgie_desktop.TabSwitcher) by replicating what Budgie WM is doing. So passing each window xid using PassItem and then openning the tab switcher using ShowSwitcher while providing the xid of the window you want to switch to. You could then close it using StopSwitcher.

Code is here:
https://github.com/BuddiesOfBudgie/budgie-desktop/blob/b7f4c06285f0b2759c404948c0ee06ad778e0c4d/src/wm/wm.vala#L1105

**Reply by erikbaender** [reply]
Thanks for the tip, but unfortunately the switcher closes by itself immediately after calling ShowSwitcher() if you don't hold down the button combination that is mapped in the settings. Is there any way around that?

**Reply by erikbaender** [reply]
I switched strategy to monitoring dbus messages and calling a script on `StopSwitcher`, but apparently that never even gets called. `PassItem` and `ShowSwitcher` show up, but `StopSwitcher` doesn't. Whats up with that?

**Reply by guillotjulien** [reply]
StopSwitcher seems to be called only when switching workspace to force stop it. Otherwise I'm seeing that the workspace switcher checks every 50ms to see if the keys are pressed, and if not, hide itself. I don't think there are any workaround for that.

**Reply by erikbaender** [reply]
Alright, thanks for your help anyways!

**Comment by erikbaender** [comment]
I've been trying to solve this in a few different ways now, but I always get stuck at the same issue:

The budgie keyboard settings don't allow for two actions to be mapped to the same key combination. This would not be problem if you could separately configure which two keys are used to operate the tab switcher, but since they're taken from this key combination in the settings it is impossible to have your own custom script launched by the same combination that operates the tab switcher.

Would you consider giving the tab switcher its own executable that performs the alt+tab action in a future release? Ideally it would receive the holding and switching key as an argument:

`/usr/bin/budgie-tab-switcher --hold-key Alt_L --switch-key Tab`

Most other budgie applications like the run dialog or the window shuffler have one or even several executables for different actions, which makes them much more useful for customizing budgie through scripts. It seems I'm forced to use a different tab switcher for now, even though I really like the budgie tab switcher.

**Reply by JoshStrobl** [reply]
> Would you consider giving the tab switcher its own executable that performs the alt+tab action in a future release?

Likely not. It needs to remain tightly coupled with the WM to facilitate the WM's ability to communicate various window states to it (mainly a list of window IDs), and having it always exist but hidden beats the situation with the launcher, which it getting rebuilt every time it is invoked. That is a flawed model and one that will be remedied in the future.

There aren't any current plans on further adjustments to the alt+tab switcher.
